addons/account/models/account_payment.py

```diff
@@ -186,585 +289,431 @@ def _get_method_codes_needing_bank_account(self):
-        if 'payment_method_id' not in rec:
-            if invoices[0].is_inbound():
-                domain = [('payment_type', '=', 'inbound')]
-            else:
-                domain = [('payment_type', '=', 'outbound')]
-            rec['payment_method_id'] = self.env['account.payment.method'].search(domain, limit=1).id
-        return rec
-
-    @api.depends('invoice_ids', 'journal_id.inbound_payment_method_ids', 'journal_id.outbound_payment_method_ids')
-    def _compute_available_payment_methods(self):
-        for p in self:
-            invoice = p.invoice_ids[:1]
-            if not invoice:
-                p.available_payment_methods = self.env['account.payment.method']
-            elif invoice.is_inbound():
-                p.available_payment_methods = self.journal_id.inbound_payment_method_ids._origin
-            else:
-                p.available_payment_methods = self.journal_id.outbound_payment_method_ids._origin
+            # Update the existing journal items.
+            # If dealing with multiple write-off lines, they are dropped and a new one is generated.
 
-    def _prepare_payment_vals(self, invoices):
-        '''Create the payment values.
+            line_ids_commands = [
+                (1, liquidity_lines.id, line_vals_list[0]),
+                (1, counterpart_lines.id, line_vals_list[1]),
+            ]
 
-        :param invoices: The invoices/bills to pay. In case of multiple
-            documents, they need to be grouped by partner, bank, journal and
-            currency.
-        :return: The payment values as a dictionary.
-        '''
-        amount = self.env['account.payment']._compute_payment_amount(invoices, invoices[0].currency_id, self.journal_id, self.payment_date)
-        values = {
-            'journal_id': self.journal_id.id,
-            'payment_method_id': self.payment_method_id.id,
-            'payment_date': self.payment_date,
-            'communication': " ".join(i.invoice_payment_ref or i.ref or i.name for i in invoices),
-            'invoice_ids': [(6, 0, invoices.ids)],
-            'payment_type': ('inbound' if amount > 0 else 'outbound'),
-            'amount': abs(amount),
-            'currency_id': invoices[0].currency_id.id,
-            'partner_id': invoices[0].commercial_partner_id.id,
-            'partner_type': MAP_INVOICE_TYPE_PARTNER_TYPE[invoices[0].move_type],
-            'partner_bank_account_id': invoices[0].invoice_partner_bank_id.id,
-        }
-        return values
+            for line in writeoff_lines:
+                line_ids_commands.append((2, line.id))
+
+            if writeoff_lines:
+                line_ids_commands.append((0, 0, line_vals_list[2]))
+
+            pay.move_id.write({
+                'partner_id': pay.partner_id.id,
+                'currency_id': pay.currency_id.id,
+                'line_ids': line_ids_commands,
+            })
```

Comment:
Rather than running 1 commands to update liquidity and counterpart, then doing 2 commands on writeoff and finally creating new lines to replace them, couldn't we, more simply, just run a 5 command and rewrite all the lines? It would look way cleaner imo.